### PR TITLE
Fix selection logic with shift on multi-click

### PR DESCRIPTION
**Pull request by Don-Vito**
## PR Checklist
* [x] Closes https://github.com/microsoft/terminal/issues/9382
* [x] CLA signed. 
* [ ] Tests added/passed
* [ ] Documentation updated. 
* [ ] Schema updated.
* [ ] I've discussed this with core contributors already. 

## Detailed Description of the Pull Request / Additional comments
The selection with shift is quite broken in 1.6.

It started with #8611 that introduces cell selection on `shift+click`.
This change resulted in the following defect:
`shift+double-click`, `shift+triple-click` select only parts  of the word.
The reason for this is that the first `shift+click` establishes the selection,
while the consequent clicks simply extend it to the relevant boundary
(aka word / line boundary)

However, the logic was broken even before #8611.
For instance, `shift+triple-click` had exactly the same handicap:
`shift+double-click` was establishing the selection and the
third click was simply extending it to the line boundary.

This PR addresses the both defects in the following manner:
upon multi-click that starts new selection we establish
a new selection on every consequent click using appropriate mode
(cell/word/line) rather than trying to extend one.
For this purpose we remember the position that started the selection.

### CODE REVIEW

**Comment by Don-Vito** [comment]
@carlos-zamora - reapplied the fix. I believe it is worth servicing.

**Comment by Don-Vito** [comment]
@zadjii-msft - please take a look at this as well. It is severe enough IMHO to go at least into 1.7.

**Review by zadjii-msft — APPROVED** [review]
Okay yea this looks right. Thanks!
\<musing>
Man, I should really think about the fact that we need TermControl tests more. With the `TermControl`->`{UwpTerminalControl, ControlInteractivity, ControlCore}` work I'm doing, this is one of those things we're definitely going to want better tests on. This mostly seems like it's interactivity logic that would be bundled in the content process. 
\</musing>

**Review by carlos-zamora — APPROVED** [review]
I wish there was a way to do this without `_lastMouseClickPosNoSelection`, but I think this is the most straightforward solution.

Tested it out manually and it feels great. Thanks!

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 1]
```diff
@@ -1342,35 +1343,55 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
+                            _terminal->SetSelectionEnd(terminalPosition, mode);
+                        }
+                        else
+                        {
+                            // If there is no selection we establish it using the selected mode
+                            // (expand both "start" and "end" selection points)
+                            _terminal->MultiClickSelection(terminalPosition, mode);
+                        }
+                        _selectionNeedsToBeCopied = true;
+                    }
```
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/TermControl.cpp:1389]: IN RETROSPECT: I don't think this is right. We might have lost the case where there's no selection, and you double/triple click on a word

@Don-Vito as an fyi

Comment (DHowett) [reply src/cascadia/TerminalControl/TermControl.cpp:1389]: I may stage a revert for this until we can get it in and get it right.

I'd like to see manual test runs for the following scenarios:

* [ ] single click = no selection
* [ ] single click and drag = selection starting from first point
* [ ] single click in unfocused pane and drag = focus pane, selection starting from first point
* [ ] double-click = selects a whole word
* [ ] triple-click = selects a whole line
* [ ] double-click and drag = selects a whole word, drag selects whole words
* [ ] triple-click and drag = selects a whole line, drag selects whole lines
* [ ] <kbd>Shift</kbd> single-click = defines start point
* [ ] second <kbd>Shift</kbd> single-click = defines end point
* [ ] <kbd>Shift</kbd> double-click = selects entire word
* [ ] <kbd>Shift</kbd> triple-click = selects entire line
* [ ] <kbd>Shift</kbd> double-click and drag = selects entire word, drag selects whole words
* [ ] <kbd>Shift</kbd> triple-click and drag = selects entire line, drag selects whole lines

**Pull request merged**